torch/distributed/fsdp/_common_utils.py

```diff
@@ -29,6 +29,24 @@
 FSDP_PREFIX = FSDP_WRAPPED_MODULE + "."
 FSDP_FLATTENED = "_fsdp_flattened"
 
+"""
+[Note: FSDP State Traversal]
+For the wrapper code path, ``_FSDPState`` is the ``FullyShardedDataParallel``
+module wrapping a fully sharded module, and for the non-wrapper code path,
+``_FSDPState`` is an object that gets embedded on a fully sharded module.
+See [Note: Fully Sharded Module] for the definition.
+
+There are three common traversal idioms: Given a root module,
+- ``_get_fsdp_states()`` returns all ``_FSDPState`` s in the tree.
+- ``get_fsdp_root_states()`` returns all local root ``_FSDPState`` s in the
+tree (i.e. those with ``_is_root == True``).
+- ``_get_fsdp_handles()``returns all ``FlatParamHandle`` s in the tree.
```

Comment:
would I be correct that the handles returned by `_get_fsdp_handles()` and `_FSDPState` returned by `_get_fsdp_states()` should always match (come from the same fully sharded module)?

Reply:
Yup. `_get_fsdp_handles` just grabs the `._handles` from the return value of `_get_fsdp_states`.